File: src/main/java/io/junseok/mealmate/domain/restaurant/service/RestaurantService.java
package io.junseok.mealmate.domain.restaurant.service;

import io.junseok.mealmate.domain.restaurant.dto.request.RestaurantRegister;
import io.junseok.mealmate.domain.restaurant.dto.response.RestaurantDetailInfo;
import io.junseok.mealmate.domain.restaurant.dto.response.RestaurantInfo;
import io.junseok.mealmate.domain.restaurant.entity.Restaurant;
import io.junseok.mealmate.domain.restaurant.repository.RestaurantRepository;
import io.junseok.mealmate.domain.restaurantmenu.dto.response.MenuInfo;
import io.junseok.mealmate.domain.restaurantmenu.repository.RestaurantMenuRepository;
import io.junseok.mealmate.exception.ErrorCode;
import io.junseok.mealmate.exception.MealMateException;
import io.junseok.mealmate.global.s3.dto.S3Response;
import io.junseok.mealmate.global.s3.service.S3UploadImage;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class RestaurantService {

    private static final Integer INITIAL_COUNT=0;
    private static final Double INITIAL_GRADE=0.0;
    private final RestaurantRepository restaurantRepository;
    private final RestaurantMenuRepository restaurantMenuRepository;
    private final S3UploadImage s3UploadImage;

    @Transactional
    public Long createRestaurant(RestaurantRegister restaurantRegister, MultipartFile file) {
        if (restaurantRepository.existsByRestaurantName(restaurantRegister.restaurantName())){
            throw new MealMateException(ErrorCode.EXIST_EMAIL);
        }
        S3Response s3Response = s3UploadImage.saveImage(file);
        Restaurant restaurant = Restaurant.builder()
            .restaurantName(restaurantRegister.restaurantName())
            .restaurantImageUrl(s3Response.imageUrl())
            .restaurantFileName(s3Response.fileName())
            .restaurantType(restaurantRegister.restaurantType())
            .likeCount(INITIAL_COUNT)
            .location(restaurantRegister.location())
            .telNum(restaurantRegister.telNum())
            .openAt(restaurantRegister.openAt())
            .closeAt(restaurantRegister.closeAt())
            .restaurantNotice(restaurantRegister.restaurantNotice())
            .grade(INITIAL_GRADE)
            .build();

        return restaurantRepository.save(restaurant).getRestaurantId();
    }

    @Transactional
    public void removeRestaurant(Long restaurantId) {
        restaurantRepository.deleteById(restaurantId);
    }

    @Transactional(readOnly = true)
    public List<RestaurantInfo> findRestaurants(String restaurantType) {
        return restaurantRepository.findAllByRestaurantType(restaurantType)
            .stream()
            .map(RestaurantInfo::fromEntity)
            .toList();
    }

    public List<RestaurantInfo> showBestRestaurants() {
        return restaurantRepository.find3TopByLikeCount()
            .stream()
            .map(RestaurantInfo::fromEntity)
            .toList();
    }

    @Transactional(readOnly = true)
    public RestaurantDetailInfo findRestaurantInfo(Long restaurantId) {
        Restaurant restaurant = restaurantRepository.findById(restaurantId)
            .orElseThrow(() -> new MealMateException(ErrorCode.EXIST_EMAIL));

        List<MenuInfo> menuInfoList = restaurantMenuRepository.findAllByRestaurant(restaurant)
            .stream()
            .map(MenuInfo::fromEntity)
            .toList();

        return RestaurantDetailInfo.builder()
            .restaurantName(restaurant.getRestaurantName())
            .restaurantImageUrl(restaurant.getRestaurantImageUrl())
            .restaurantType(restaurant.getRestaurantType())
            .restaurantNotice(restaurant.getRestaurantNotice())
            .likeCount(restaurant.getLikeCount())
            .grade(restaurant.getGrade())
            .openAt(restaurant.getOpenAt())
            .closeAt(restaurant.getCloseAt())
            .menuList(menuInfoList)
            .build();
    }
}

File: src/main/java/io/junseok/mealmate/domain/restaurant/service/RestaurantService.kt
package io.junseok.mealmate.domain.restaurant.service

import io.junseok.mealmate.domain.restaurant.dto.request.RestaurantRegister
import io.junseok.mealmate.domain.restaurant.dto.response.RestaurantDetailInfo
import io.junseok.mealmate.domain.restaurant.dto.response.RestaurantInfo
import io.junseok.mealmate.domain.restaurant.dto.response.toCreateRestaurantResponse
import io.junseok.mealmate.domain.restaurant.entity.Restaurant
import io.junseok.mealmate.domain.restaurant.repository.QueryRestaurantRepository
import io.junseok.mealmate.domain.restaurant.repository.RestaurantRepository
import io.junseok.mealmate.domain.restaurantmenu.dto.response.MenuInfo
import io.junseok.mealmate.domain.restaurantmenu.dto.response.toCreateMenuInfo
import io.junseok.mealmate.domain.restaurantmenu.entity.RestaurantMenu
import io.junseok.mealmate.domain.restaurantmenu.repository.RestaurantMenuRepository
import io.junseok.mealmate.exception.ErrorCode
import io.junseok.mealmate.exception.MealMateException
import io.junseok.mealmate.global.s3.service.S3UploadImage
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile


@Service
class RestaurantService(
    private val restaurantRepository: RestaurantRepository,
    private val restaurantMenuRepository: RestaurantMenuRepository,
    private val queryRestaurantRepository: QueryRestaurantRepository,
    private val s3UploadImage: S3UploadImage
) {
    @Transactional
    fun createRestaurant(restaurantRegister: RestaurantRegister, file: MultipartFile): Long? {
        if (restaurantRepository.existsByRestaurantName(restaurantRegister.restaurantName)) {
            throw MealMateException(ErrorCode.EXIST_EMAIL)
        }
        val s3Response =  s3UploadImage.saveImage(file)
        val restaurant = Restaurant(
            restaurantName = restaurantRegister.restaurantName,
            restaurantImageUrl = s3Response.imageUrl,
            restaurantFileName = s3Response.fileName,
            restaurantType = restaurantRegister.restaurantType,
            likeCount = INITIAL_COUNT,
            location = restaurantRegister.location,
            telNum = restaurantRegister.telNum,
            openAt = restaurantRegister.openAt,
            closeAt = restaurantRegister.closeAt,
            restaurantNotice = restaurantRegister.restaurantNotice,
            grade = INITIAL_GRADE
        )
        return restaurantRepository.save(restaurant).restaurantId
    }

    @Transactional
    fun removeRestaurant(restaurantId: Long) {
        restaurantRepository.deleteById(restaurantId)
    }

    /** TODO
     * Change TO nullable
     */
    @Transactional(readOnly = true)
    fun findRestaurants(restaurantType: String?): List<RestaurantInfo>? {
        return restaurantType.let { queryRestaurantRepository.findAllRestaurant(restaurantType!!)
            ?.map { it.toCreateRestaurantResponse()!! } } ?: restaurantRepository.findAll()
                .map { it.toCreateRestaurantResponse()!! }
    }

    fun showBestRestaurants(): List<RestaurantInfo> =
        restaurantRepository.findTop3ByOrderByLikeCountDesc()
            .map { it.toCreateRestaurantResponse()!! }

    @Transactional(readOnly = true)
    fun findRestaurantInfo(restaurantId: Long): RestaurantDetailInfo {
        val restaurant = restaurantRepository.findByIdOrNull(restaurantId)
            ?: throw MealMateException(ErrorCode.EXIST_EMAIL)

        val menuInfoList = restaurantMenuRepository.findAllByRestaurant(restaurant)
            .map { it.toCreateMenuInfo() }

        return RestaurantDetailInfo(
            restaurantName = restaurant.restaurantName,
            restaurantImageUrl = restaurant.restaurantImageUrl,
            restaurantType = restaurant.restaurantType,
            restaurantNotice = restaurant.restaurantNotice,
            restaurantTelNum = restaurant.telNum,
            location = restaurant.location,
            likeCount = restaurant.likeCount,
            grade = restaurant.grade,
            openAt = restaurant.openAt,
            closeAt = restaurant.closeAt,
            menuList = menuInfoList
        )
    }

    fun findByRestaurantId(restaurantId: Long): Restaurant =
        restaurantRepository.findById(restaurantId)
            .orElseThrow { MealMateException(ErrorCode.NOT_EXIST_RESTAURANT) }

    @Transactional(readOnly = true)
    fun findRestaurantByName(restaurantName: String): RestaurantInfo {
        return restaurantRepository.findByRestaurantName(restaurantName)?.toCreateRestaurantResponse()
            ?: throw MealMateException(ErrorCode.NOT_EXIST_RESTAURANT)
    }

    companion object {
        private const val INITIAL_COUNT = 0
        private const val INITIAL_GRADE = 0.0
    }
}
